File: 152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int x = nums[0];
        int maxx = 1;
        int minn = 1;
        int t;
        for(int i = 0; i < nums.size(); i++){
            t = minn;
            minn = min(nums[i], min(nums[i]*minn, maxx * nums[i]));
            maxx = max(nums[i], max(nums[i]*t, maxx * nums[i]));
            x = max(x, maxx);
        }
        return x;
    }
};

File: Python/152.py
class Solution:
    def f(self,cp, sol, nums, index, dp):
        if index >= len(nums):
            return
        
        if (index, cp) in dp:
            return

        if cp * nums[index] > sol[0]:
            sol[0] = cp * nums[index]
        
        dp[(index, cp)] = 1
        self.f(cp*nums[index], sol, nums, index+1, dp)
        if nums[index]<= 0:
            self.f(1, sol, nums, index+1, dp)

    def maxProduct(self, nums: List[int]) -> int:
        dp = {}
        sol = [float('-inf')]
        _ = self.f(1, sol, nums, 0, dp)
        return sol[0]
